### PR TITLE
Build Script Fixes & Improvements

### PR DESCRIPTION
**Pull request by RPangrle**
* Fix build script to work with Xcode 9 code coverage
* Disable shallow analyze during build
* Switch to 'build-for-testing' and 'test-without-build' to eliminate duplicative builds during the ios and mac framework steps
* Boot the iOS Simulator at the top of the build script if it will be needed as part of the script run, to reduce the delay when starting iOS tests
* Split travis build into three separate jobs to allow parallel builds

    The boot up time for a travis worker is pretty significant (> 2 min) so spinning up workers that take less then two minutes gets somewhat wasteful if each job ends up running in serial. However being able to parallelize our builds can result in faster overall build times if we can get enough build workers simultaneously. Additional only having to restart a subportion of the build for random build failures can come in handy. I ended up splitting them based on what targets can potentially reuse products from previous targets to try to minimize how much gets built duplicatively, while still parallelizing our builds.

* Added time elapsed printing in the summary for each target
* Added time elapsed printing for the whole script at the end

### CODE REVIEW

**Comment by unpluggedk** [comment]
From the result, this is fabulous and I have nothing against it.
I might need some time looking at the actual script before approving it. Can someone take a look too?
Thanks

**Comment by oldalton** [comment]
@unpluggedk I'm on the same page, I'd need some more time understanding what it does before I can do a meaningful review

File: build.py [hunk 1]
```diff
@@ -146,7 +160,12 @@ def xcodebuild_command(self, operation, xcpretty) :
 		
 		command += " -scheme \"" + self.scheme + "\" -configuration " + default_config
 		
-		if (operation == "test" and "codecov" in self.operations) :
+		# The shallow analyzer is buggy. Stupidly buggy, causing random failures that didn't fail the build on things like
+		# headers not being found. If Apple can't make this reliable then we should short circuit it out of our build
```
Comment (oldalton) [inline build.py]: :)

File: build.py [hunk 2]
```diff
@@ -292,17 +332,40 @@ def do_operation(self, operation) :
 		
 		print 
 		return exit_code
+	
+	def requires_simulator(self) :
+		if self.platform is not "iOS" :
+			return False
+		if "test" in self.operations :
```
Comment (oldalton) [inline build.py]: this makes sense 👍

**Pull request merged**